### ISSUE TITLE
[FORM] This value is not valid. (code 1dafa156-89e1-4736-b832-419c2e501fca)

### ISSUE THREAD

**Issue by shakaran** [symfony/symfony#37898]
**Symfony version(s) affected**: 5.0.1

**Description**  

I am getting the weird error description with that hash, with is not very friend human error. I try to google and [look at the source](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Validator/Constraints/Form.php#L21) at vendor/symfony/form/Extension/Validator/Constraints/Form.php, which I only found a Form contraint called `NOT_SYNCHRONIZED_ERROR` but no documentation or explanation in docs about which is doing or which cause the origen of this error.

I had a old legacy code inherit from other developer that was using:

```
$form->handleRequest($request);
$form->submit($request);
```
At the same time. So submit method probably is not syncronized and raise that error in isValid() form. Should be catched and displayed as global error form in a more friendly way.

**footer.html.twig**
```html
<div class="col-xl-4 col-md-6 col-lg-4">
						{{ render(controller('App\\Controller\\FooterNewsletterController::renderNewsletterForm', {'base' : 1})) }}
					</div>
```

**footer_newsletter/index.html.twig**
```html
<div class="footer_widget">
							<h3 class="footer_title">Subscribirse</h3>
							<form id="suscriber_footer_form" name="footer_newsletter_type" class="newsletter_form" 
								action="{{ url('newsletter_subscription') }}">
								<input name="email_susbscriber" id="email_susbscriber" type="text" placeholder="Introduce tu email">
								<button name="send_subscriber" id="send_subscriber" type="submit" value="Enviar">Enviar</button>
							</form>
							<p class="newsletter_text">No usaremos tu correo para enviar spam.</p>
						</div>

```
**src/Form/FooterNewsletterType.php**
```php
<?php

namespace App\Form;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\OptionsResolver\OptionsResolver;
use Symfony\Component\Validator\Constraints\Email;
use Symfony\Component\Validator\Constraints\NotBlank;
use Symfony\Component\Form\Extension\Core\Type\EmailType;
use Symfony\Component\Form\Extension\Core\Type\SubmitType;

class FooterNewsletterType extends AbstractType
{
    public const EMAIL_FIELD_NAME    = 'email_susbscriber';
    
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
            $builder->add(self::EMAIL_FIELD_NAME, EmailType::class, [
                'required' => true,
                'empty_data'  => '',
                'constraints' => [
                    new NotBlank(),
                    new Email(['mode' => 'strict']),
                ],
            ]);
            $builder->add('send_subscriber', SubmitType::class);
            
            $builder->setMethod(Request::METHOD_POST);
    }

    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults([
            'allow_extra_fields' => true,
            'error_bubbling' => false,
        ]);
    }
    
    public function getBlockPrefix()
    {
        return 'footer_newsletter_type';
    }
}
```
**src/Controller/FooterNewsletterController.php**
```php
<?php

namespace App\Controller;

use App\Form\FooterNewsletterType;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\HttpFoundation\JsonResponse;

class FooterNewsletterController extends AbstractController
{
    /**
     * @Route("/footer/newsletter", name="footer_newsletter")
     */
    public function renderNewsletterForm(Request $request): Response
    {
        $data = [];
        
        return $this->render('footer_newsletter/index.html.twig', [
            $data
        ]);
    }
    
    /**
     * 
     *
     * @Route("/newsletter/suscripcion", name="newsletter_subscription", methods={"POST"})
     */
    public function newsletterSubscribed(Request $request): Response
    {
        $form = $this->createForm(FooterNewsletterType::class, null, ['csrf_protection' => false])->handleRequest($request);
        
        $form->handleRequest($request);
        $form->submit($request); // PROBLEM HERE (second time processed)
            
        if ($request->isMethod('POST') && $request->isXmlHttpRequest()) 
        {

            if ($form->isSubmitted() && $form->isValid())
            {
                [FooterNewsletterType::EMAIL_FIELD_NAME => $email] = $form->getData();
                
                if (empty($email))
                {
                    // Not a spam, save email, etc...
                    //$org = (new Organization())->setName($email);
                    
                    $this->addFlash('success', sprintf('Thank you "%s"! See you later on 🐝️!', 'bla')); /
                }
                else
                {
                    // Spam detected!
                    $warning = sprintf('🐛 SPAM detected: email: "%s", honeypot content: "%s"', $email, 'blo');
                    $this->addFlash('warning', $warning);
                }
                
                return new JsonResponse('Form valid ' . print_r($form->getData(), TRUE), 200);
            }
            else
            {
                $errors = [];
                foreach ($this->getErrors(true) as $error)
                 {

 var_dump((string)$error->getMessage());
var_dump((string)$error->getCause());
var_dump((string)$error->getOrigin());

var_dump((string)$error->getMessageTemplate());
var_dump((string)$error->getMessageParameters());


}

var_dump((string)($this->getErrors(true)));exit;

                $response = new JsonResponse('', 200);
                $response->setData('errors', print_r($errors, TRUE));
                return $response;
            }
            
        }
        
        
    }
}
```

Javascript
```js
if($( "#suscriber_footer_form" ))
		{
        		$( "#suscriber_footer_form" ).submit(function( event ) 
                {	
                	  event.preventDefault();

                	  var form = $('#suscriber_footer_form')[0]; // You need to use standard javascript object here
                	  var formData = new FormData(form);
                	  
                	  $.ajax({
                        type: "POST",
                        url: $(this).attr("action"),
                        data: formData + '&send_subscriber=true',
                        processData: false,
    					contentType: false,
    					cache: false,
                        success: (data) => 
                        {
                        	console.log('Submission was successful.');
                			console.log(data);
                
                            $('#popup-box-content').html('<p style="text-align: center">Thank you for ' +
                            'subscribing to The Polyglot Developer newsletter!</p>');
                        },
                        error: function (data) {
                            console.log('An error occurred.');
                            console.log(data);
                        },
                    });
                	  
                	  
                	  
                });
         }
```

**Output:**
```
string(24) "This value is not valid."
string(109) "Object(Symfony\Component\Form\Form):
    This value is not valid. (code 1dafa156-89e1-4736-b832-419c2e501fca)"

```
**Possible Solution**  
Detect when submit is called after handleRequest or improve error messaging to be more friendly

Additionally. Using only:

`$form->submit($request);`

Produces the error also of no syncronized.

Using only:

`$form->handleRequest($request);`

Will detect the form as NO subbmited.

And using:

```php
$form->handleRequest($request);
$form->submit($request);
```

Will produce no syncronized.

At some point the form send via ajax is not picking the values after form submition.

**Comment by shakaran** [comment]
I am debugging more with the profiler, and it seems that the global form is failing for this validation:

![image](https://user-images.githubusercontent.com/14254/90838674-c381a400-e355-11ea-8624-949442bde797.png)

Which shows in form "Compound forms expect an array or NULL on submission":

![image](https://user-images.githubusercontent.com/14254/90838702-d6947400-e355-11ea-9e9e-716362828bf5.png)

Pretty weird that I cannot get via the getErrors(true) the global error right, pretty confusing, since it is triggering the general message value of 'invalid_message' in the resolver.

I try put in the resolver:

`'compound'   => false,`

And additionaly:

```
'data_class' => null,
 'empty_data'   => '',
```

As per [comment in this issue](https://github.com/dustin10/VichUploaderBundle/issues/769#issuecomment-458533267)

Then I get the error:

`Uncaught PHP Exception Symfony\Component\Form\Exception\LogicException: "You cannot add children to a simple form. Maybe you should set the option "compound" to true?" at /vendor/symfony/form/Form.php line 870 `

**Comment by shakaran** [comment]
I manage to figure out which is exactly the problem and solve it.

The exactly problem cames assuming that I can pass a $request object in submit method and symfony will handle it. Since it is expecting an array, and the data comes via request content payload, I have to call to $request->getContent() and get a string that it is converted later to array. All this, because $form->handleRequest didn't understand the data that comes from the ajax request payload data and fill the form structure data.

So basically change:

`$form->submit($request);`

To:
   ```php
     $data = [];
        parse_str($request->getContent(), $data);
  
        $form->submit($data); //($request);
        $form->handleRequest($request);
```

So, even with that, the error get in form general should be more explicit or detect at some point that the user don't make weird things like I did, or at least print a message more understable to solve the real problem

**Comment by xabbuh** [comment]
I am not sure we can do much better than what we see in the screenshots in https://github.com/symfony/symfony/issues/37898#issuecomment-677974234. The `ConstraintViolation` object holds all the information why the form is not valid (it expects an array or `null`, but the passed data is an object of type `Request`).

**Comment by xabbuh** [comment]
I am closing here as I don't think we can do much about this. But please feel free to reply with some ideas on how to improve and we can consider to reopen.
